Feature/friendship management

charbelraffoul opened the pull request:
# Feature: Comprehensive Friend Request Management System

## Overview

This pull request introduces a comprehensive and robust friend request management system to our application, enhancing both backend and frontend components. By upgrading the `UserProfile` model, refining repository layers, enhancing the ViewModel, improving the user interface, and expanding our testing suite, we provide users with seamless capabilities to send, accept, decline, cancel, and delete friend requests. These enhancements not only improve user interaction but also ensure data consistency, integrity, and reliability across the platform.

## Changes

### 1. Backend Enhancements

#### a. `UserProfile.kt` Enhancements

- **Added Fields:**
  - **`sentReq`**: A list of user IDs to whom the current user has sent friend requests.
  - **`recReq`**: A list of user IDs who have sent friend requests to the current user.

  These additions enable tracking of both outgoing and incoming friend requests, facilitating better management and display of friend-related interactions within the application.

#### b. `UserProfileRepository` Updates

- **Comprehensive Documentation:**
  - Added detailed documentation for all existing and newly created functions to enhance code readability and maintainability.

- **New Methods Implemented:**
  - **`sendFriendRequest`**: Allows a user to send a friend request to another user.
  - **`acceptFriendRequest`**: Enables a user to accept a received friend request, establishing a mutual friendship.
  - **`declineFriendRequest`**: Allows a user to decline a received friend request.
  - **`deleteFriend`**: Facilitates the removal of an existing friendship between two users.
  - **`cancelFriendRequest`**: Enables a user to cancel a previously sent friend request.
  - **`getRecReqProfiles` & `getSentReqProfiles`**: Fetches profiles based on received and sent friend request IDs, respectively.

  These methods ensure a full spectrum of friend request operations, enhancing social interactions within the app.

#### c. `UserProfileRepositoryFireStore` Enhancements

- **`deleteUserProfile` Update:**
  - Enhanced to ensure that when a user deletes their profile, the corresponding friendship entries are removed from both the user's and their friends' profiles, maintaining data integrity.

- **Implementation of Friend Request Methods:**
  - All new repository methods are implemented using Firestore transactions to guarantee atomic operations, ensuring that friend request actions are executed reliably and consistently.

- **Optimized Data Retrieval:**
  - Methods for fetching received and sent friend request profiles have been optimized to handle empty states gracefully and efficiently retrieve necessary user profiles based on their IDs.

### 2. `UserProfileViewModel` Enhancements

- **State Management:**
  - Introduced `MutableStateFlow` properties for managing received and sent friend request profiles, allowing for reactive updates and better state handling within the UI.
  - **Imports Added:**
    - `import kotlinx.coroutines.flow.asStateFlow`

- **Data Fetching Functions:**
  - Added functions to fetch received and sent friend request profiles from the repository and update the respective state flows. These functions handle successful data retrieval and error scenarios, ensuring the ViewModel remains in sync with repository data.

- **Implemented Methods:**
  - **`acceptFriend`**: Accepts a friend request from a specified user, updating both the friends list and received requests accordingly.
  - **`declineFriendRequest`**: Declines a friend request from a specified user, removing the request from the received requests list.
  - **`sendRequest`**: Sends a friend request to a specified user, updating the sent requests list.
  - **`cancelFriendRequest`**: Cancels a previously sent friend request, removing it from the sent requests list.
  - **`deleteFriend`**: Updated to handle the removal of a friend, ensuring the friendship is deleted from both users' profiles.

  These methods provide comprehensive management of friend requests, enhancing the overall user experience by enabling seamless social interactions.

### 3. Frontend Enhancements

#### a. `FriendsScreen` Enhancements

- **Friend Requests List:**
  - **Added Feature:** Introduced a new list displaying incoming friend requests that are pending user response.
  - **Expandable List:** Allows users to show or hide the `recReq` list as needed.
  - **Integrated Scrolling:** Both the existing friends list and the new friend requests list are scrollable together, ensuring a cohesive user experience.
  - **Actionable Requests:** Users can accept or decline friend requests directly from the list.
    - **On Acceptance:** Accepted requests appear in the friends list and are removed from `recReq`.
    - **On Decline:** Declined requests remove the requester from `recReq` and remove the current user from the sender's `sentReq`.

- **Search Bar Improvements:**
  - **UI Enhancements:** Improved the search bar for better usability and aesthetics.
  - **Functionality Fixes:** Resolved issues related to search functionality to ensure accurate and efficient filtering of friends and friend requests.

#### b. `AddFriends` Screen Enhancements

- **Sent Friend Requests List:**
  - **Added Feature:** Introduced a new list displaying users to whom the current user has sent friend requests that are pending response.
  - **Expandable List:** Allows users to manage their sent requests effectively.
  - **Scrollable and Filterable:** Integrated with search functionality for easy navigation and filtering.
  - **Actionable Requests:** Users can cancel sent friend requests directly from the list, providing greater control over their outgoing requests.

- **Handling Mutual Friend Requests:**
  - **Alert Dialog:** Triggers when a user sends a friend request to someone who has already sent them one.
  - **User Decisions:** Provides options to:
    - **Accept:** Establish the friendship.
    - **Decide Later:** Defer the decision.
    - **Reject:** Decline the friend request.

  This feature ensures smooth handling of reciprocal friend requests, enhancing user experience and interaction.

### 4. Testing Enhancements

#### a. Test Suite Refinement

- **Removed Outdated Tests:**
  - **`deleteFriend should remove friend from friends list`**: Removed due to evolved function behavior.
  - **`addFriend should add friend to friends list`**: Similarly removed as it no longer aligns with the current implementation.

  These tests were obsolete and did not correspond to the updated functionalities, ensuring the test suite remains relevant and accurate.

#### b. Added Comprehensive Tests

- **Increased Test Coverage:**
  - Added numerous new tests covering the newly implemented friend request functionalities, achieving an impressive **80.0% line coverage**, especially considering the challenges associated with testing Firestore interactions.

- **Key Tests Implemented:**
  - **`sendFriendRequest`**: Ensures sending a friend request updates both the sender's `sentReq` and the receiver's `recReq`.
  - **`acceptFriendRequest`**: Verifies that accepting a friend request establishes a mutual friendship and updates request lists appropriately.
  - **`declineFriendRequest`**: Confirms that declining a request removes the requester from `recReq` and the user from the sender's `sentReq`.
  - **`cancelFriendRequest`**: Checks that cancelling a sent request removes it from both `sentReq` and `recReq`.
  - **`deleteFriend`**: Validates that deleting a friend removes the friendship from both users' `friends` lists.
  - **`updateLoginStreak`**: Ensures the user's login streak updates correctly based on their last login date.
  - **`getRecReqProfiles` & `getSentReqProfiles`**: Tests that fetched profiles based on request IDs are accurate and reflect the current state.

- **Testing Firestore Interactions:**
  - Utilized mocking techniques to simulate Firestore transactions and document snapshots, ensuring each friend request operation behaves as expected under various scenarios.
  - Implemented comprehensive error handling within tests to verify that failure callbacks are invoked appropriately when operations do not succeed.

- **UI Tests Integration:**
  - Enhanced UI tests to interact with new frontend components, ensuring that user actions such as sending, accepting, declining, and cancelling friend requests are correctly reflected in both the UI and backend data.

#### c. Code Duplication

- **New Code Duplication:**
  - Introduced **3.9%** new code duplication in the test suite.

- **Reason for Duplication:**
  - To effectively mock Firestore actions for each list (`friends`, `recReq`, and `sentReq`), similar functions had to be defined for each. A generalized function approach was ineffective during the mocking process, necessitating duplication to ensure accurate and reliable test behaviors.

- **Future Considerations:**
  - Investigate advanced mocking frameworks or refactor repository methods to reduce duplication while maintaining testability.
  - Explore parameterized tests or utility functions to minimize repetitive code without compromising clarity and effectiveness.

#### d. Enhanced Testability for Reviewers

- **Prepopulated Friend Requests:**
  - Sent friend requests to all users from my profile. This setup allows reviewers to fully test and interact with the newly implemented features within this PR, ensuring comprehensive validation and a seamless review process.

  By populating the sent requests, reviewers can experience the complete flow of sending, accepting, declining, and cancelling friend requests, providing a practical and thorough evaluation of the feature's functionality.

### 5. Backend Enhancements (Continued)

- **Firestore Transactions:**
  - All friend request operations (`send`, `accept`, `decline`, `cancel`, `delete`) are executed within Firestore transactions, ensuring atomicity and maintaining data consistency even in cases of concurrent modifications or failures.

- **Error Handling:**
  - Implemented comprehensive error handling across all methods with meaningful exception messages and logging. This facilitates easier debugging and provides clear feedback to users in case of failures.

## Testing

### 1. Unit Tests

- **Repository Methods:**
  - Added and updated unit tests for all new repository methods to ensure functionality and handle edge cases effectively.

- **ViewModel Functions:**
  - Tested the new data fetching functions to verify proper state updates and error handling.

### 2. Integration Tests

- **Firestore Transactions:**
  - Verified Firestore transactions for friend request operations to ensure data integrity and atomicity.

- **Frontend-Backend Interaction:**
  - Ensured seamless communication between frontend components and backend services, validating data consistency across different scenarios.

### 3. UI Tests

- **Friend Request Operations:**
  - Simulated user actions for sending, accepting, declining, and cancelling friend requests to validate end-to-end functionality and ensure a smooth user experience.

- **Expandable Lists:**
  - Tested the expand/collapse functionality of friend requests lists to ensure they behave as expected and maintain UI consistency.

- **Alert Dialogs:**
  - Verified that alert dialogs appear correctly when handling mutual friend requests and that user decisions (accept, decline, decide later) are processed appropriately.

---

Thank you for reviewing this comprehensive update! Your feedback and suggestions are highly appreciated to make this feature even better.

MathisHage commented on app/src/main/java/com/github/se/orator/model/profile/UserProfileRepositoryFirestore.kt:
```diff
@@ -340,6 +350,372 @@ class UserProfileRepositoryFirestore(private val db: FirebaseFirestore) : UserPr
+          onSuccess(friends)
+        }
+        .addOnFailureListener { exception ->
+          Log.e("UserProfileRepository", "Error fetching friends profiles", exception)
```
Maybe replacing "friends profiles" by "pending received requests" would describe the problem more precisely.

```suggestion
          Log.e("UserProfileRepository", "Error fetching pending received requests", exception)
```

MathisHage commented on app/src/main/java/com/github/se/orator/model/profile/UserProfileRepositoryFirestore.kt:
```diff
@@ -340,6 +350,372 @@ class UserProfileRepositoryFirestore(private val db: FirebaseFirestore) : UserPr
+          onSuccess(friends)
+        }
+        .addOnFailureListener { exception ->
+          Log.e("UserProfileRepository", "Error fetching friends profiles", exception)
```
Maybe replacing "friends profiles" by "pending sent requests" would describe the problem more precisely.

```suggestion
          Log.e("UserProfileRepository", "Error fetching pending sent requests", exception)
```

MathisHage commented on app/src/main/java/com/github/se/orator/model/profile/UserProfileRepositoryFirestore.kt:
```diff
@@ -340,6 +350,372 @@ class UserProfileRepositoryFirestore(private val db: FirebaseFirestore) : UserPr
+  }
+
+  /**
+   * Get profiles the user received requests from based on the UIDs stored in the user's profile.
+   *
+   * @param recReqUIds List of UIDs of the users the user received requests from to be retrieved.
+   * @param onSuccess Callback to be invoked with the list of friends' profiles.
+   * @param onFailure Callback to be invoked on failure with the exception.
+   */
+  override fun getSentReqProfiles(
+      friendUids: List<String>,
+      onSuccess: (List<UserProfile>) -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
```
Again, mismatch between the name of the first parameter in the documentation and in the actual code. `sentReqUIds` would probably be a more suited name.

MathisHage commented on app/src/main/java/com/github/se/orator/model/profile/UserProfileRepositoryFirestore.kt:
```diff
@@ -340,6 +350,372 @@ class UserProfileRepositoryFirestore(private val db: FirebaseFirestore) : UserPr
+   * @param onSuccess Callback to be invoked with the list of friends' profiles.
+   * @param onFailure Callback to be invoked on failure with the exception.
+   */
+  override fun getRecReqProfiles(
```
This method is essentially the same as `getFriendsProfiles`. To avoid code duplication, I would suggest creating a common method with the shared code.

charbelraffoul replied:
This is what I initially had done, but when mocking to be able to test, I wanted each method to do different things. This is why I did this

MathisHage commented on app/src/main/java/com/github/se/orator/model/profile/UserProfileRepositoryFirestore.kt:
```diff
@@ -340,6 +350,372 @@ class UserProfileRepositoryFirestore(private val db: FirebaseFirestore) : UserPr
+   * @param onSuccess Callback to be invoked with the list of friends' profiles.
+   * @param onFailure Callback to be invoked on failure with the exception.
+   */
+  override fun getSentReqProfiles(
```
Again, this method is essentially the same as `getFriendsProfiles`. To avoid code duplication, I would suggest creating a common method with the shared code.

MathisHage commented on app/src/main/java/com/github/se/orator/model/profile/UserProfileRepositoryFirestore.kt:
```diff
@@ -340,6 +350,372 @@ class UserProfileRepositoryFirestore(private val db: FirebaseFirestore) : UserPr
+
+          // Optionally, check if the friend has already sent a request to the current user
+          if (friendRecReq.contains(currentUid)) {
+            throw Exception("The user has already sent you a friend request.")
```
Maybe it would be more logical to then accept the person's request instead of just failing. But this is just my point of view, not an issue.

charbelraffoul replied:
This will never be the case as I am handling it with the AlertDialog, so we will never have a case where we call sendFriendRequest and the user has already received a friend request from the same profile. I removed this condition.

MathisHage commented on app/src/main/java/com/github/se/orator/model/profile/UserProfileRepositoryFirestore.kt:
```diff
@@ -340,6 +350,372 @@ class UserProfileRepositoryFirestore(private val db: FirebaseFirestore) : UserPr
+          }
+
+          if (!friendSentReq.contains(currentUid)) {
+            throw Exception("No sent friend request from current user to this user.")
```
```suggestion
            throw Exception("No sent friend request from distant user to this user.")
```

MathisHage commented on app/src/main/java/com/github/se/orator/model/profile/UserProfileRepositoryFirestore.kt:
The code quality is very good, but I just find that there is a lot of code duplication. This is the case for the methods `sendFriendRequest`, `acceptFriendRequest`, `declineFriendRequest`, `cancelFriendRequest` and `deleteFriend`, for which the code to retrieve the documents and add the listeners could be put in a helper function, instead of being re-written 5 times.

MathisHage replied:
The code duplication is now under the required threshold, so I will not insist further on these recommendation. But changing this would have been better for a more efficient code maintainability.

MathisHage replied:
The code duplication is now under the required threshold, so I will not insist further on these recommendation. But changing this would have been better for a more efficient code maintainability.

charbelraffoul replied:
Thank you for your feedback, Mathis! I have taken your suggestion into consideration and implemented a helper function to reduce the code duplication in the sendFriendRequest, acceptFriendRequest, declineFriendRequest, cancelFriendRequest, and deleteFriend methods. The helper function handles the common logic for retrieving user documents and executing the transaction, improving maintainability and reducing redundancy. Let me know if you'd like me to share more details or the updated implementation!

MathisHage replied:
Nicely done ! Thanks for taking this into account.

MathisHage commented on app/src/main/java/com/github/se/orator/ui/friends/AddFriends.kt:
```diff
@@ -184,60 +228,197 @@ fun AddFriendsScreen(
+                    Text(
+                        text = sentRequest.bio ?: "No bio available",
+                        style = MaterialTheme.typography.bodySmall,
+                        color = AppColors.secondaryTextColor,
```
The colors used should come from `MaterialTheme.colorScheme` to be linked with the centralized color manager.
```suggestion
                        color = MaterialTheme.colorScheme.secondary,
```

MathisHage commented on app/src/main/java/com/github/se/orator/ui/friends/AddFriends.kt:
```diff
@@ -184,60 +228,197 @@ fun AddFriendsScreen(
-                    .show()
-              },
+              .testTag("userItem#${user.uid}"),
       color = AppColors.LightPurpleGrey,
```
Same remark for the color use.

MathisHage commented on app/src/main/java/com/github/se/orator/ui/friends/AddFriends.kt:
```diff
@@ -184,60 +228,197 @@ fun AddFriendsScreen(
+                    Text(
+                        text = user.bio ?: "No bio available",
+                        style = MaterialTheme.typography.bodySmall,
+                        color = AppColors.secondaryTextColor,
```
Same remark for the color use.

MathisHage commented on app/src/main/java/com/github/se/orator/ui/friends/FriendsScreen.kt:
```diff
@@ -373,3 +473,75 @@ fun currentFriendStreak(lastLoginDateString: String?, currentStreak: Long): Long
+                Text(
+                    text = friendRequest.bio ?: "No bio available",
+                    style = MaterialTheme.typography.bodySmall,
+                    color = AppColors.secondaryTextColor,
```
The color should be retrieved using `MaterialTheme.colorScheme` as discussed earlier.

MathisHage commented on app/src/test/java/com/github/se/orator/model/profile/UserProfileRepositoryFirestoreTest.kt:
I think that all the mocked objects should be initialized in the function annotated with `@Before` to avoid code duplication.

charbelraffoul replied:
The code duplication for UserProfileRepositoryFirestoreTest is 0.0%, the code duplication is caused by code in the file UserProdileRepositoryFirestore. This is mainly because of the methods getRecReqProfiles and getSentReqProfiles, which I implmeneted to te able to test.
![image](https://github.com/user-attachments/assets/cea4f460-0cbc-422b-9f05-3b1b20b171a5)

MathisHage replied:
Well, i am assuming that sonar cloud disregards analysis on test files, because lines 379 to 392 are used 5 times in the tests that are added in the file `UserProfileRepositoryFirestoreTest.kt`. I think this is enough repetition to at least make the local string variables into class variables.

MathisHage replied:
Thank you for taking this into account !

MathisHage reviewed (changes requested):
This PR adds great changes to the app. The code quality is in my opinion good overall, but some minor issues such as code duplication could sometimes be better handled. The use of the color is also not done using the chosen color manager (`MaterialTheme.colorScheme`).

Also, this PR introduces a lot of changes. Maybe too much. I think that breaking this down into 4-5 smaller PR's would have allowed a better organization of the tasks, as well as giving a break to the poor reviewer (myself :wink:).

charbelraffoul commented on app/src/main/java/com/github/se/orator/model/profile/UserProfileViewModel.kt:
```diff
@@ -149,27 +196,157 @@ class UserProfileViewModel(internal val repository: UserProfileRepository) : Vie
+
+            // Update local state for friends
+            val updatedFriendsList =
+                userProfile_.value?.friends?.toMutableList()?.apply { add(friend.uid) }
```
Thanks for this suggestion! Using plus indeed feels cleaner and avoids unnecessary conversion to a mutable list. I'll make the change as suggested.

charbelraffoul commented on app/src/main/java/com/github/se/orator/model/profile/UserProfileRepositoryFirestore.kt:
```diff
@@ -340,6 +350,372 @@ class UserProfileRepositoryFirestore(private val db: FirebaseFirestore) : UserPr
         }
   }
 
+  /**
+   * Get profiles the user received requests from based on the UIDs stored in the user's profile.
+   *
+   * @param recReqUIds List of UIDs of the users the user received requests from to be retrieved.
+   * @param onSuccess Callback to be invoked with the list of friends' profiles.
+   * @param onFailure Callback to be invoked on failure with the exception.
+   */
+  override fun getRecReqProfiles(
+      friendUids: List<String>,
+      onSuccess: (List<UserProfile>) -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
```
You are right, thanks!

charbelraffoul commented:
Hi @MathisHage,

I hope you're doing well.

I wanted to provide you with an update on the recent changes based on your valuable suggestions:

## Summary of Changes

1. **Implemented All Suggestions**
   - Carefully reviewed and addressed all the suggestions you provided.
   - Modified the necessary parts of the code to enhance functionality and maintainability.

2. **Refactored `getProfiles` Methods**
   - Introduced a **helper function** for the three `getProfiles` methods:
     - `getFriendsProfiles`
     - `getRecReqProfiles`
     - `getSentReqProfiles`
   - This refactoring has significantly reduced code duplication to 2.0%, which satisfies the required condition, and streamlined the codebase.

3. **Improved Code Coverage**
   - Increased our code coverage to **82.5%**.
   - Enhanced reliability and robustness of the code through comprehensive testing.

## Impact of Changes

- **Reduced Code Duplication:** The helper function eliminates repetitive code, making the codebase cleaner and easier to maintain.
- **Enhanced Maintainability:** With centralized logic in the helper function, future modifications can be made more efficiently.
- **Better Reliability:** Higher code coverage ensures that more parts of the code are tested, reducing the likelihood of bugs.

## Next Steps

- **Review:** Could you please take a moment to review the changes at your convenience?
- **Feedback:** I would greatly appreciate any further feedback or questions you might have to continue improving the project.

Thank you for your continuous support and guidance!

MathisHage commented on app/src/main/java/com/github/se/orator/model/profile/UserProfileViewModel.kt:
```diff
@@ -265,33 +442,41 @@ class UserProfileViewModel(internal val repository: UserProfileRepository) : Vie
+
+            // Update the sent requests list by removing the friend
+            val updatedFriends =
+                userProfile_.value?.friends?.toMutableList()?.apply { remove(friend.uid) }
```
Hey, what do you think about this suggestion ?

MathisHage reviewed (changes requested):
The suggestions have been taken into account very seriously. Thank you about that ! There is still one suggestion that remains unanswered though, and another one about code duplication that I will disregard as the code duplication is rather low, and it does not constitute a major problem in my opinion.

charbelraffoul commented:
Replying to "Hey, what do you think about this suggestion (using kotlin collections minus method) ?". This was actually in the main in the method deleteFriend (which is why I did all comparable actions such as adding and removing elements to lists like that). But I will fix it now.

charbelraffoul commented:
I have refactored the tests to align with the updated implementation of the UserProfileRepository. This includes adjustments to ensure compatibility with the refactored methods such as sendFriendRequest, acceptFriendRequest, declineFriendRequest, cancelFriendRequest, and deleteFriend. Proper mocking of Firestore transactions and document references has been implemented, and the interactions with Firestore have been thoroughly verified to cover both successful and failure scenarios. I believe all requested changes have been approached, ensuring that the tests comprehensively validate the updated repository's behavior.

charbelraffoul commented:
The duplication on the new code is now 0.0% and the line coverage is 83.6% percent. Thanks for your feedbacks that helped us ensure high code quality and maintability!

MathisHage reviewed (approved):
The requested changes are now officially all addressed, well done ! The code's quality is really good, which will benefit the rest of the development of the app in a really positive way ! Well done !

charbelraffoul commented:
Thank you for all the valuable suggestions and feedback throughout this process. Your insights were instrumental in helping me improve the quality and robustness of the code. I truly appreciate the time and effort you invested in providing detailed reviews, which not only enhanced this feature but also contributed positively to the overall development of the project. Looking forward to continuing this collaborative effort!
